lib/attrsets.nix

```diff
@@ -494,4 +494,26 @@ rec {
   zip = builtins.trace
     "lib.zip is deprecated, use lib.zipAttrsWith instead" zipAttrsWith;
 
+  /* Return the cartesian product of attribute set value combinations.
+
+    Example:
+      cartesianProductOfSets { a = [ 1 2 ]; b = [ 10 20 ]; }
+      => [
+           { a = 1; b = 10; }
+           { a = 1; b = 20; }
+           { a = 2; b = 10; }
+           { a = 2; b = 20; }
+         ]
+  */
+  cartesianProductOfSets = let
+    crossMap = set: attrName: attrValues: let
+        f = attrValue: (set // { "${attrName}" = attrValue; });
+      in builtins.map f attrValues;
+    crossApply = attrName: attrValues: list:
+      builtins.concatMap (set: crossMap set attrName attrValues) list;
+    crossSet = set: let
+        names = builtins.attrNames set;
+        f = (list: name: crossApply name (set."${name}") list);
+      in builtins.foldl' f [{}] names;
+  in crossSet;
```

Comment:
I feel like the `let in` are a bit distracting, how about
```nix
{
  cartesianProductOfSets = attrsOfLists:
    lib.foldl' (listOfAttrs: attrName:
      concatMap (attrs:
        map (listValue: attrs // { ${attrName} = listValue; }) attrsOfLists.${attrName}
      ) listOfAttrs
    ) [{}] (attrNames attrsOfLists);
}
```

Instead

Reply:
you're right, looks much cleaner. i incorporated your suggestion.